warehouse/accounts/forms.py

```diff
@@ -415,7 +415,19 @@ class TOTPAuthenticationForm(TOTPValueMixin, _TwoFactorAuthenticationForm):
     def validate_totp_value(self, field):
         totp_value = field.data.replace(" ", "").encode("utf8")
 
-        if not self.user_service.check_totp_value(self.user_id, totp_value):
+        try:
+            self.user_service.check_totp_value(self.user_id, totp_value)
+        except otp.OutOfSyncTOTPError:
+            user = self.user_service.get_user(self.user_id)
+            user.record_event(
+                tag=EventTag.Account.LoginFailure,
+                request=self.request,
+                additional={"reason": "out_of_sync_totp"},
+            )
+            raise wtforms.validators.ValidationError(
+                _("Invalid TOTP code. Your device time may be out of sync.")
+            )
+        except otp.InvalidTOTPError:
```

Comment:
I like this, but is this form used during OTP provisioning? I think it's only for authentication post-provisioning.

Reply:
aa40787